util/oteltest/src/oteltest/__init__.py

```diff
@@ -0,0 +1,172 @@
+        script was running and the output of the script (stdout, stderr, returncode).
+        """
+
+
+@dataclasses.dataclass
+class Request:
+    request: dict
+    headers: dict
+    test_elapsed_ms: int
+
+    def get_header(self, name):
+        return self.headers.get(name)
+
+    def to_json(self):
+        return json.dumps(self.to_dict())
+
+    def to_dict(self):
+        return {
+            "request": self.request,
+            "headers": self.headers,
+            "test_elapsed_ms": self.test_elapsed_ms,
+        }
+
+
+class Telemetry:
+    def __init__(
+        self,
+        log_reqs: Optional[List[Request]] = None,
+        metric_reqs: Optional[List[Request]] = None,
+        trace_reqs: Optional[List[Request]] = None,
+    ):
+        self.log_reqs: List[Request] = log_reqs or []
+        self.metric_reqs: List[Request] = metric_reqs or []
+        self.trace_reqs: List[Request] = trace_reqs or []
+
+    def add_log(self, log: dict, headers: dict, test_elapsed_ms: int):
+        self.log_reqs.append(Request(log, headers, test_elapsed_ms))
+
+    def add_metric(self, metric: dict, headers: dict, test_elapsed_ms: int):
+        self.metric_reqs.append(Request(metric, headers, test_elapsed_ms))
+
+    def add_trace(self, trace: dict, headers: dict, test_elapsed_ms: int):
+        self.trace_reqs.append(Request(trace, headers, test_elapsed_ms))
+
+    def get_trace_requests(self) -> List[Request]:
+        return self.trace_reqs
+
+    def num_metrics(self) -> int:
+        out = 0
+        for req in self.metric_reqs:
+            for rm in req.request["resourceMetrics"]:
+                for sm in rm["scopeMetrics"]:
+                    out += len(sm["metrics"])
+        return out
+
+    def metric_names(self) -> set:
+        out = set()
+        for req in self.metric_reqs:
+            for rm in req.request["resourceMetrics"]:
+                for sm in rm["scopeMetrics"]:
+                    for metric in sm["metrics"]:
+                        out.add(metric["name"])
+        return out
+
+    def num_spans(self) -> int:
+        out = 0
+        for req in self.trace_reqs:
+            for rs in req.request["resourceSpans"]:
+                for ss in rs["scopeSpans"]:
+                    out += len(ss["spans"])
+        return out
+
+    def has_trace_header(self, key, expected) -> bool:
+        for req in self.trace_reqs:
+            actual = req.get_header(key)
+            if expected == actual:
+                return True
+        return False
+
+    def to_json(self):
+        return json.dumps(self.to_dict(), indent=2)
+
+    def to_dict(self):
+        return {
+            "log_reqs": [req.to_dict() for req in self.log_reqs],
+            "metric_reqs": [req.to_dict() for req in self.metric_reqs],
+            "trace_reqs": [req.to_dict() for req in self.trace_reqs],
+        }
+
+    def __str__(self):
+        return self.to_json()
+
+
+def trace_attribute_as_str_array(tr: dict, attr_name) -> [str]:
```

Comment:
Since the readme/examples don't showcase this and it isn't in private, could we have a quick docstring usage?

Reply:
Yeah that function deserves a docstring, but it turns out that nothing is actually using it, so it's been removed for now. There's a ton of testing capabilities around extracting and comparing telemetry that will be useful for testing later but I'm thinking keep this PR as small as possible.

Reply:
Ah!! Sounds good.